From 19e9141cfc036569e0b3ebf39d127cd5d112190f Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 19 Jun 2025 11:30:57 +0800
Subject: [PATCH 210/236] mtk: hostapd: fix hostapd crash due to missing
 bss->flink assignment

It seems that the code for finding new bss->flink during removing link
is missing. Add it back to prevent hostapd crash due to use-after-free.

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/drivers/driver_nl80211.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index cd7bb00b4..659cb10eb 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10019,7 +10019,7 @@ int nl80211_remove_link(struct i802_bss *bss, int link_id)
 	struct wpa_driver_nl80211_data *drv = bss->drv;
 	struct i802_link *link;
 	struct nl_msg *msg;
-	int ret;
+	int ret, i;
 	u8 link_addr[ETH_ALEN];
 
 	wpa_printf(MSG_DEBUG, "nl80211: Remove link (ifindex=%d link_id=%u)",
@@ -10041,6 +10041,14 @@ int nl80211_remove_link(struct i802_bss *bss, int link_id)
 	nl80211_update_active_links(bss, link_id);
 	bss->valid_links &= ~BIT(link_id);
 
+	/* Choose new deflink if we are removing that link */
+	if (bss->flink == link) {
+		for_each_link_default(bss->valid_links, i, 0) {
+			bss->flink = &bss->links[i];
+			break;
+		}
+	}
+
 	/* If this was the last link, reset default link */
 	if (!bss->valid_links) {
 		/* TODO: Does keeping freq/bandwidth make sense? */
-- 
2.45.2

